core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala

```diff
@@ -39,13 +39,18 @@ object SparkTransportConf {
    */
   def fromSparkConf(_conf: SparkConf, module: String, numUsableCores: Int = 0): TransportConf = {
     val conf = _conf.clone
-
-    // Specify thread configuration based on our JVM's allocation of cores (rather than necessarily
-    // assuming we have all the machine's cores).
-    // NB: Only set if serverThreads/clientThreads not already set.
+    val executorId = conf.get("spark.executor.id", "")
+    val isDriver = executorId == SparkContext.DRIVER_IDENTIFIER ||
+          executorId == SparkContext.LEGACY_DRIVER_IDENTIFIER
+    val role = if (isDriver) "driver" else "executor"
```

Comment:
This is wrong. A quick search for `SparkTransportConf.fromSparkConf` shows this being called by `ExternalShuffleService`, which is neither driver not executor.

Reply:
@vanzin , The configuration is read not only from "role", but also "module" in this enhancement. If there is no configurations for "role", it will fallback to "module". For ExternalShuffleService, we can just ignore "role" and use "module" of "shuffle", such as, spark.shuffle.*.  This enhancement weighs on the bottleneck of driver when they are too many RPC messages due to large number of small tasks. 
Of course, we can have more "roles" if there are areas can be performance improved.

Reply:
@vanzin , I get your point. Actually, @jerryshao also pointed out the similar problem and told me to introduce "role" to make code clear. At the time, I tried to make the "role" as explicit argument to this method. However, it involves in a lot of code changes in many areas to get right "role".  It's not a simple task to be covered in this PR.  It seems that they don't want this PR to be that complex. Maybe we can do it when "role" is clearly defined in worker, master, driver etc.

Reply:
You can introduce role with an empty default value. Then anything that is not a driver or executor would not define it, and would get the old behavior.

Reply:
@vanzin , ok, let me try. thanks.

Reply:
@vanzin , I just changed code to fix the potential issue of daemon miss-using executor's configuration. Each executor has executor id being set to spark.executor.id which gives me a chance to judge whether it's executor or daemons. The default role is set to None and only RPCEnv tries to set role if there are driver or executor specific configurations. thanks.